README


This archive contains the distribution WebService-GlucoseBuddy,
version 1.113420:

  Interface to a glucosebuddy.com account

This software is copyright (c) 2011 by Pete Smith.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.




Changes
Revision history for WebService-GlucoseBuddy

1.113420  2011-12-08 23:25:38 Europe/London
    Add from and to date options to search logs
    Improve documentation

1.113400  2011-12-06 16:13:55 Europe/London
    Initial release


lib/WebService/GlucoseBuddy.pm
package WebService::GlucoseBuddy;
# ABSTRACT: Interface to a glucosebuddy.com account

use Moose 1.24;
use namespace::autoclean 0.13;
use MooseX::Iterator 0.11;
use WWW::Mechanize 1.70;
use Readonly 1.03;
use Text::CSV 1.21;
use DateTime::Format::Strptime 1.5;
use Carp 1.20;

use WebService::GlucoseBuddy::Log;
use WebService::GlucoseBuddy::Log::Reading;

Readonly my $SERVICE_URI    => 'https://www.glucosebuddy.com';
Readonly my $DT_FORMAT      => '%m/%d/%Y %T';


has username => (
    is          => 'ro',
    isa         => 'Str',
    required    => 1,
);


has password => (
    is          => 'ro',
    isa         => 'Str',
    required    => 1,
);

has _mech => (
    is          => 'ro',
    isa         => 'WWW::Mechanize',
    lazy_build  => 1,
);

sub _build__mech {
    my $self = shift;

    my $mech = WWW::Mechanize->new;
    $mech->get($SERVICE_URI . '/login');

    unless ($mech->success) {
        carp 'Could not connect to glucosebuddy.com';
    }

    $mech->submit_form(
        with_fields => {
            login       => $self->username,
            password    => $self->password,
        }
    );

    unless ($mech->uri->path eq '/logs/new') {
        carp 'Log in failed';
    }

    return $mech;
}

has _logs => (
    is          => 'ro',
    isa         => 'ArrayRef',
    lazy_build  => 1,
);

{
    my $dt_formatter = DateTime::Format::Strptime->new(
        pattern     => $DT_FORMAT,
        on_error    => 'croak',
    );

    sub _build__logs {
        my $self = shift;

        my $mech = $self->_mech;
        $mech->get('/logs/MyExportedGlucoseBuddyLogs.csv');

        my $logs_file = $mech->content;
        open my $logs_fh, '<', \$logs_file;

        my $csv = Text::CSV->new;

        my @logs;

        # throw away header
        <$logs_fh>;

        while (my $row = $csv->getline($logs_fh)) {
            # change time column to DT object so we can filter it later
            $row->[5] = $dt_formatter->parse_datetime($row->[5]);

            push @logs, $row;
        }

        return \@logs;
    }
}


sub logs {
    my $self = shift;
    my %args = @_;
    
    my @logs = @{ $self->_logs };

    if ($args{from}) {
        @logs = grep { $_->[5] >= $args{from} } @logs;
    }

    if ($args{to}) {
        @logs = grep { $_->[5] <= $args{to} } @logs;
    }

    my @log_objects;

    for (@logs) {
        my $reading = WebService::GlucoseBuddy::Log::Reading->new(
            type    => $_->[0],
            value   => $_->[1],
            unit    => $_->[2],
        );

        push @log_objects => WebService::GlucoseBuddy::Log->new(
            reading => $reading,
            time    => $_->[5],
            name    => $_->[3],
            event   => $_->[4],
            notes   => $_->[6],
        );
    }

    return MooseX::Iterator::Array->new(collection => \@log_objects);
}

__PACKAGE__->meta->make_immutable;

1;

__END__
=pod

=head1 NAME

WebService::GlucoseBuddy - Interface to a glucosebuddy.com account

=head1 VERSION

version 1.113420

=head1 METHODS

=head2 logs [%args]

Returns an L<MooseX::Iterator> iterator of L<WebService::GlucoseBuddy::Log> objects, each 
representing a log

Args can contain:

=over 1

=item from

A L<DateTime> object to search logs on or after this time

=item to

A L<DateTime> object to search logs on or before this time

=back

=head1 CONSTRUCTOR

=head2 new

Constructor requires these arguments:

=over 1

=item username

Your glucosebuddy.com username

=item password

Your glucosebuddy.com password

=back

=head1 AUTHOR

Pete Smith <pete@cubabit.net>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2011 by Pete Smith.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/WebService/GlucoseBuddy/Log.pm
package WebService::GlucoseBuddy::Log;
# ABSTRACT: A log from a glucosebuddy logfile

use Moose 1.24;
use namespace::autoclean 0.13;


has reading => (
    is  => 'ro',
    isa => 'WebService::GlucoseBuddy::Log::Reading',
);


has name => (
    is  => 'ro',
    isa => 'Str',
);


has event => (
    is  => 'ro',
    isa => 'Str',
);


has time => (
    is  => 'ro',
    isa => 'DateTime',
);


has notes => (
    is  => 'ro',
    isa => 'Str',
);

__PACKAGE__->meta->make_immutable;

1;


__END__
=pod

=head1 NAME

WebService::GlucoseBuddy::Log - A log from a glucosebuddy logfile

=head1 VERSION

version 1.113420

=head1 ATTRIBUTES

=head2 reading

A L<WebService::GlucoseBuddy::Log::Reading> object for the reading

=head2 name

The name given for the log entry

=head2 event

The event name for the log entry

=head2 time

A L<DateTime> object for the time of the reading. This has a floating timezone as glucosebuddy.com
does not provide one.

=head2 notes

Notes for the log entry

=head1 AUTHOR

Pete Smith <pete@cubabit.net>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2011 by Pete Smith.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/WebService/GlucoseBuddy/Log/Reading.pm
package WebService::GlucoseBuddy::Log::Reading;
# ABSTRACT: A reading from a glucosebuddy log

use Moose 1.24;
use namespace::autoclean 0.13;


has type => (
    is  => 'ro',
    isa => 'Str',
);


has value => (
    is  => 'ro',
    isa => 'Num',
);


has unit => (
    is  => 'ro',
    isa => 'Str',
);

__PACKAGE__->meta->make_immutable;

1;


__END__
=pod

=head1 NAME

WebService::GlucoseBuddy::Log::Reading - A reading from a glucosebuddy log

=head1 VERSION

version 1.113420

=head1 ATTRIBUTES

=head2 type

The type of reading

=head2 value

Teh value of the reading

=head2 unit

The units used for the reading

=head1 AUTHOR

Pete Smith <pete@cubabit.net>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2011 by Pete Smith.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



Makefile.PL

use strict;
use warnings;



use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Interface to a glucosebuddy.com account",
  "AUTHOR" => "Pete Smith <pete\@cubabit.net>",
  "BUILD_REQUIRES" => {
    "DateTime" => "0.70",
    "Test::MockObject" => "1.20110612",
    "Test::More" => "0.98",
    "URI" => "1.59",
    "strict" => 0,
    "warnings" => 0
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "WebService-GlucoseBuddy",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "WebService::GlucoseBuddy",
  "PREREQ_PM" => {
    "Carp" => "1.20",
    "DateTime::Format::Strptime" => "1.5",
    "Moose" => "1.24",
    "MooseX::Iterator" => "0.11",
    "Readonly" => "1.03",
    "Text::CSV" => "1.21",
    "WWW::Mechanize" => "1.70",
    "namespace::autoclean" => "0.13"
  },
  "VERSION" => "1.113420",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



